core/src/filesystem/media/analyze_media_job/mod.rs

```diff
@@ -152,93 +163,26 @@ impl JobExt for AnalyzeMediaJob {
 		task: Self::Task,
 	) -> Result<JobTaskOutput<Self>, JobError> {
 		let mut output = Self::Output::default();
+		let mut subtasks = Vec::new();
 
 		match task {
-			AnalyzeMediaTask::AnalyzeImage(id) => {
-				// Get media by id from the database
-				let media_item: Media = ctx
-					.db
-					.media()
-					.find_unique(media::id::equals(id.clone()))
-					.with(media::metadata::fetch())
-					.exec()
-					.await
-					.map_err(|e: prisma_client_rust::QueryError| {
-						JobError::TaskFailed(e.to_string())
-					})?
-					.ok_or_else(|| {
-						JobError::TaskFailed(format!(
-							"Unable to find media item with id: {}",
-							id
-						))
-					})?
-					.into();
-
-				let path = media_item.path;
-				let page_count = get_page_count(&path, &ctx.config)?;
-				output.images_analyzed += 1;
-
-				// Check if a metadata update is neded
-				if let Some(metadata) = media_item.metadata {
-					// Great, there's already metadata!
-					// Check if the value matches the currently recorded one, update if not.
-					if let Some(meta_page_count) = metadata.page_count {
-						if meta_page_count != page_count {
-							ctx.db
-								.media_metadata()
-								.update(
-									media_metadata::media_id::equals(media_item.id),
-									vec![media_metadata::page_count::set(Some(
-										page_count,
-									))],
-								)
-								.exec()
-								.await?;
-							output.media_updated += 1;
-						}
-					} else {
-						// Page count was `None` so we update it.
-						ctx.db
-							.media_metadata()
-							.update(
-								media_metadata::id::equals(media_item.id),
-								vec![media_metadata::page_count::set(Some(page_count))],
-							)
-							.exec()
-							.await?;
-						output.media_updated += 1;
-					}
-				} else {
-					// Metadata doesn't exist, create it
-					let new_metadata = ctx
-						.db
-						.media_metadata()
-						.create(vec![
-							media_metadata::id::set(media_item.id.clone()),
-							media_metadata::page_count::set(Some(page_count)),
-						])
-						.exec()
-						.await?;
-
-					// And link it to the media item
-					ctx.db
-						.media()
-						.update(
-							media::id::equals(media_item.id),
-							vec![media::metadata::connect(media_metadata::id::equals(
-								new_metadata.id,
-							))],
-						)
-						.exec()
-						.await?;
-					output.media_updated += 1;
-				}
+			AnalyzeMediaTask::UpdatePageCount(id) => {
+				task_page_count::execute(id, ctx, &mut output).await?
+			},
+			AnalyzeMediaTask::AnalyzePageDimensions(id) => {
+				task_analyze_dimensions::execute(id, ctx, &mut output).await?
+			},
+			AnalyzeMediaTask::FullAnalysis(id) => {
+				// First page count needs to be updated
+				task_page_count::execute(id.clone(), ctx, &mut output).await?;
+				// Then we can queue a dimensions analysis job
+				subtasks.push(AnalyzeMediaTask::AnalyzePageDimensions(id));
 			},
```

Comment:
This is my solution for the need to run the page count analysis before the page dimensions analysis (because page dimensions relies on having page count).

I think a superior solution that is outside the scope of this pull request would be proving some sort of dependency graph for job tasks that enforces scheduling dependent tasks to occur first.

Reply:
Elaborating on this, this has one undesirable effect on user experience: when the task starts. As an illustrative example, imagine a user has a library with 100 items in it, and they select "Analyze Media." This fires the AnalyzeMedia job whose `init` queues up one `AnalyzeMediaTask::FullAnalysis` for each of the 100 items in the library.

This causes the indicator on the web UI to display `0/100`. However, when the first task runs (executing the page count logic) and queues up `AnalyzeMediaTask::AnalyzePageDimensions`, it's going to update the UI to read `1/101` and so on until it slowly ticks up to `200/200`. This is very annoying, and also confusing ("why does my 100 item library count to 200 when analyzing?" "why did an update 2 months later make it go to 300?").

I'm going to refactor to make `AnalyzeMediaTask::FullAnalysis` do both tasks' logic in series, which is fine because it looks like tasks run sequentially in the current job system anyway. This will make the UI count up to the number of media items being analyzed. It might be worth exploring task parallelism in the future.

Reply:
I know I was overall iffy about the UX when I realized that how I structured the job/scanner rewrite makes it easy for a growing queue, and I called it out in the docs so people aren't too confused: https://www.stumpapp.dev/guides/basics/scanner#walk-series

I think the performance gain during a scan is worth the cut to UX slightly, since calculating _all_ tasks at the beginning of a scan would be potentially very expensive, but I agree that ideally we should aim to reduce that sort of behavior where we can

> I'm going to refactor to make AnalyzeMediaTask::FullAnalysis do both tasks' logic in series, which is fine because it looks like tasks run sequentially in the current job system anyway

I think this works 👍 down the road, it might be good to explore an opt-in matrix sort of approach where I check various analysis tasks which the job understands at the start.

They do run sequentially because of my below point:

> It might be worth exploring task parallelism in the future

If they can be structured in a way to not issue DB writes, tasks can be parallelized. Throughout Stump's lifetime, I have observed and tested that Prisma+SQLite does *not* play well with concurrent writes (even with WAL enabled). Sometimes I wish I chose something like Postgres out the gate so folks can run 2+ scans concurrently without worry, or even just proper parallelism within a scan, but that can be a consideration down the road 😅